### PR TITLE
[BEAM-6504] Create Portable sideInput handler for Dataflow (Part One)

### PR DESCRIPTION
**Pull request by HuangLED**
This PR creates a SideInput handler inside dataflow fnApi runner. 

There will be another PR (in-progress) to actually wire this independent module into ProcessRemoteBundleOperation. 

With these two PRs, we have majority of validateRunner sideinput tests passing. 

------------------------

Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

It will help us expedite review of your Pull Request if you tag someone (e.g. `@username`) to look at it.

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | --- | --- | --- | ---
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_Verify/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/) </br> [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | --- | --- | --- | ---

### CODE REVIEW

**Comment by HuangLED** [comment]
+ R: @rohdesamuel

**Comment by HuangLED** [comment]
Update. The Part two PR has passed all 28 integration tests in ViewTest.java.

File: runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/fn/control/DataflowSideInputHandlerFactory.java [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+import org.apache.beam.runners.fnexecution.state.StateRequestHandlers;
+import org.apache.beam.sdk.coders.Coder;
+import org.apache.beam.sdk.coders.KvCoder;
+import org.apache.beam.sdk.transforms.Materializations;
+import org.apache.beam.sdk.transforms.windowing.BoundedWindow;
+import org.apache.beam.sdk.values.PCollectionView;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class DataflowSideInputHandlerFactory
```
Comment (angoenka) [inline runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/fn/control/DataflowSideInputHandlerFactory.java]: Java doc for what we are doing in this file.

File: runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/fn/control/DataflowSideInputHandlerFactory.java [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+  }
+
+  @Override
+  public <T, V, W extends BoundedWindow> StateRequestHandlers.SideInputHandler<V, W> forSideInput(
+      String pTransformId,
+      String sideInputId,
+      RunnerApi.FunctionSpec accessPattern,
+      Coder<T> elementCoder,
+      Coder<W> windowCoder) {
+    checkArgument(pTransformId != null, String.format("Expect a valid PTransform ID."));
```
Comment (angoenka) [inline runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/fn/control/DataflowSideInputHandlerFactory.java]: Lets check for both null and empty string

File: runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/fn/control/DataflowSideInputHandlerFactory.java [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.runners.dataflow.worker.fn.control;
+
+import static com.google.common.base.Preconditions.checkState;
```
Comment (angoenka) [inline runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/fn/control/DataflowSideInputHandlerFactory.java]: Import fix to org.apache.beam.vendor.guava.v20_0

**Comment by angoenka** [comment]
Thanks @HuangLED 
Few minor comments.

**Comment by HuangLED** [comment]
Comments addressed. PTAL. 

Thanks for reviewing!

File: runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/fn/control/DataflowSideInputHandlerFactory.java [hunk 4]
```diff
@@ -65,7 +66,9 @@ private DataflowSideInputHandlerFactory(
       RunnerApi.FunctionSpec accessPattern,
       Coder<T> elementCoder,
       Coder<W> windowCoder) {
-    checkArgument(pTransformId != null, String.format("Expect a valid PTransform ID."));
+    checkArgument(
+        pTransformId != null && pTransformId.length() > 0,
+        String.format("Expect a valid PTransform ID."));
```
Comment (angoenka) [inline runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/fn/control/DataflowSideInputHandlerFactory.java]: We don't need `String.format` as we are not substituting any parameters.

**Comment by swegner** [comment]
Just spotless failing now. I'll submit a formatting fix and merge once the tests are green.

**Pull request merged**